packages/perseus/src/util/react-render.tsx

```diff
@@ -2,6 +2,23 @@ import {RenderStateRoot} from "@khanacademy/wonder-blocks-core";
 import * as React from "react";
 import * as ReactDOMClient from "react-dom/client";
 
+// We keep a cache of elements that have already had createRoot() called on
+// them. If we don't do this, React 18 warns every subsequent call:
+//   > You are calling ReactDOMClient.createRoot() on a container that has
+//   > already been passed to createRoot() before.
+//
+// Note that this cache uses a WeakMap<> so that the cache isn't causing a
+// memory leak (by retaining elements after they're released everywhere else).
+const reactRootCache = new WeakMap<HTMLElement, ReactDOMClient.Root>();
```

Comment:
ACHTUNG! I'd appreciate some eyes on these changes just to make sure I've not done something bad. :)

Reply:
One question: what is the behavior in the current code (with React 16) if `render` is called multiple times with a stateful component? Is the state preserved across renders? If it isn't, then I think caching the React roots will change the behavior (the state of rendered components will now be preserved).

Reply:
The deprecated `render()` [documentation](https://react.dev/reference/react-dom/render#caveats) says that multiple calls with the same host DOM node will maintain state (at least that's how I read it):

> If you call render on the same domNode more than once, React will update the DOM as necessary to reflect the latest JSX you passed. React will decide which parts of the DOM can be reused and which need to be recreated by [“matching it up”](https://react.dev/learn/preserving-and-resetting-state) with the previously rendered tree. Calling render on the same domNode again is similar to calling the [set function](https://react.dev/reference/react/useState#setstate) on the root component: React avoids unnecessary DOM updates.